llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -7424,6 +7424,19 @@ SDValue RISCVTargetLowering::lowerINSERT_VECTOR_ELT(SDValue Op,
     Vec = convertToScalableVector(ContainerVT, Vec, DAG, Subtarget);
   }
 
+  MVT OrigContainerVT = ContainerVT;
+  SDValue OrigVec = Vec;
+  // If we know the index we're going to insert at, we can shrink Vec so that
+  // we're performing the scalar inserts and slideup on a smaller LMUL.
+  if (auto *CIdx = dyn_cast<ConstantSDNode>(Idx)) {
+    if (auto ShrunkVT = getSmallestVTForIndex(ContainerVT, CIdx->getZExtValue(),
+                                              DL, DAG, Subtarget)) {
+      ContainerVT = *ShrunkVT;
+      Vec = DAG.getNode(ISD::EXTRACT_SUBVECTOR, DL, ContainerVT, Vec,
```

Comment:
You also need to update VecVT here.  That drives the VL chosen below, and leads to the bug in the test that I had previously pointed out.  Having VL=8 with ContainerVT=m1 is not a legal combination.  You're very lucky that nothing else is breaking, as the

Reply:
I think part of this comment might have gotten cut off?

Reply:
I think there's a gap in my understanding, the AVL here is only used for the `vmv.s.x`, which only cares if it's zero or greater than zero. And at VLEN=128, choosing AVL=8 at LMUL=1 SEW=64 will set VL to something > 0?

Reply:
See https://github.com/llvm/llvm-project/pull/66087#discussion_r1339046777, we're discussing the same issue in two different parts of the review.